File: Data Structures & Algorithms/max-water-container/submission-0.java
class Solution {
    public int maxArea(int[] heights) {
        int n = heights.length;
        int left = 0;
        int right = n-1;
        int maxArea = 0;
        while (left < right) {
            int h = Math.min(heights[left], heights[right]);
            int newArea = (right-left)*h;
            maxArea = Math.max(maxArea, newArea);

            if (heights[left] < heights[right]) {
                left++;
            } else {
                right--;
            }
        }
        return maxArea;
    }
}

File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-0.js
class Solution {
    /**
     * @param {number[]} heights
     * @return {number}
     */
    largestRectangleArea(heights) {
      const idxStack = [] // stack of index have no-desc heights
      let maxArea = 0
      const h = [...heights, 0] // add sentinel value, avoid mutate input

      for (let i = 0; i < h.length; i++) {
        // if current h < top stack height
        //  -> it's the right corner of rectanle
        //  -> remove all higher histogram to calc Area
        while (
          idxStack.length > 0 && 
          h[idxStack.at(-1)] > h[i]
        ) {
          const height = h[idxStack.pop()]
          const width = idxStack.length > 0 
          ? i - idxStack.at(-1) - 1
          : i
          const newArea = width * height
          maxArea = Math.max(maxArea, newArea)
        }

        idxStack.push(i)
      }

      return maxArea
    }
}
